measure_optimize.py

```diff
@@ -2,180 +2,2412 @@
-            self.Sigma = error_cov
+        if upper_diagonal_only:
+            m.total_fim = pyo.Var(m.dim_fim_half, initialize=init_fim)
+        else:
+            m.total_fim = pyo.Var(m.dim_fim, m.dim_fim, initialize=init_fim)
+
+        ### compute FIM
+        def eval_fim(m, a, b):
+            """
+            Add constraints to calculate FIM from unit contributions
+            FIM = sum(cov_y[i,j]*unit FIM[i,j]) for all i, j in n_responses
+
+            a, b: indices for FIM, iterate in parameter set
+            """
+            if a <= b:
+                summi = 0
+                for i in m.n_responses:
+                    for j in m.n_responses:
+                        # large_idx, small_idx are needed because cov_y is only defined the upper triangle matrix
+                        # the FIM order is i*num_measurement + j no matter if i is the smaller one or the bigger one
+                        large_idx = max(i, j)
+                        small_idx = min(i, j)
+                        summi += (
+                            m.cov_y[small_idx, large_idx]
+                            * self.unit_fims[i * self.num_measure_dynamic_flatten + j][
+                                a
+                            ][b]
+                        )
+
+                # if diagonal elements, a small element can be added to avoid rank deficiency
+                if a == b:
+                    return m.total_fim[a, b] == summi + m.fim_diagonal_small_element
+                # if not diagonal, no need to add small number
+                else:
+                    return m.total_fim[a, b] == summi
+            # FIM is symmetric so no need to compute again
+            else:
+                return m.total_fim[a, b] == m.total_fim[b, a]
+
+        def integer_cut_0(m):
+            """Compute the total number of measurements and time points selected
+            This is for the inequality constraint of integer cut.
+            """
+            return m.total_number_measurements == sum(
+                m.cov_y[i, i] for i in range(self.num_measure_dynamic_flatten)
+            )
+
+        def integer_cut_0_ineq(m):
+            """Ensure that at least one measurement or time point is selected
+            integer cut that cuts the solution of all 0
+            """
+            return m.total_number_measurements >= 1
+
+        def total_dynamic(m):
+            """compute the total number of time points from DCMs are selected
+            This is for the inequality constraint of total number of time points from DCMs < total number of measurements limit
+            """
+            return m.total_number_dynamic_measurements == sum(
+                m.cov_y[i, i]
+                for i in range(
+                    self.n_static_measurements, self.num_measure_dynamic_flatten
+                )
+            )
+
+        ### cov_y constraints
+        def y_covy1(m, a, b):
```

Comment:
Please add comments mapping all constraints back to equations in the manuscript

Reply:
add Eq. number in manuscript

Reply:
added for all equations